core/commands/IssueRelationshipAddCommand.php

```diff
@@ -0,0 +1,206 @@
+	/**
+	 * The relationship type id
+	 *
+	 * @var integer
+	 */
+	private $typeId;
+
+	/**
+	 * The source issue
+	 *
+	 * @var BugData
+	 */
+	private $sourceIssue = null;
+
+	/**
+	 * The target issue
+	 *
+	 * @var BugData
+	 */
+	private $targetIssue = null;
+
+	/**
+	 * Constructor
+	 *
+	 * @param array $p_data The command data.
+	 */
+	function __construct( array $p_data ) {
+		parent::__construct( $p_data );
+	}
+
+	/**
+	 * Validate the data.
+	 */
+	function validate() {
+		$t_type = $this->payload( 'type', array( 'id' => BUG_RELATED ) );
+		$this->typeId = $this->getRelationTypeId( $t_type );
+
+		$t_source_issue_id = helper_parse_issue_id( $this->query( 'issue_id' ) );
+		$t_target_issue_ref = $this->payload( 'issue' );
+
+		if( !isset( $t_target_issue_ref['id'] ) ) {
+			throw new ClientException(
+				'Invalid issue id',
+				ERROR_INVALID_FIELD_VALUE,
+				array( 'issue_id' )
+			);
+		}
+
+		$t_target_issue_id = helper_parse_issue_id( $t_target_issue_ref['id'], 'target_issue_id' );
+
+		$this->sourceIssue = bug_get( $t_source_issue_id, true );
+
+		$t_update_threshold = config_get( 'update_bug_threshold', null, null, $this->sourceIssue->project_id );
+
+		# Ensure user has access to update the source issue
+		if( !access_has_bug_level( $t_update_threshold, $t_source_issue_id ) ) {
+			throw new ClientException(
+				'Access denied to add relationship',
+				ERROR_ACCESS_DENIED
+			);
+		}
+
+		# Ensure that source and target issues are not the same
+		if( $t_source_issue_id == $t_target_issue_id ) {
+			throw new ClientException(
+				"Issue can't have relationship to itself",
+				ERROR_RELATIONSHIP_SAME_BUG
+			);
+		}
+
+		# Ensure that related issue exists and gets its information
+		$this->targetIssue = bug_get( $t_target_issue_id, true );
+
+		# Ensure source issue is not read-only
+		if( bug_is_readonly( $t_source_issue_id ) ) {
+			throw new ClientException(
+				sprintf( "Issue %d is read-only", $t_source_issue_id ),
+				ERROR_BUG_READ_ONLY_ACTION_DENIED,
+				array( $t_source_issue_id )
+			);
+		}
+
+		# Ensure that user can view target issue
+		$t_view_threshold = config_get( 'view_bug_threshold', null, null, $this->targetIssue->project_id );
+		if( !access_has_bug_level( $t_view_threshold, $t_target_issue_id ) ) {
+			throw new ClientException(
+				sprintf( "Issue %d is read-only", $t_target_issue_id ),
+				ERROR_RELATIONSHIP_ACCESS_LEVEL_TO_DEST_BUG_TOO_LOW,
+				array( $t_target_issue_id )
+			);
+		}
+	}
+
+	/**
+	 * Process the command.
+	 *
+	 * @returns array Command response
+	 */
+	protected function process() {
+		if( $this->sourceIssue->project_id != helper_get_current_project() ) {
+			# in case the current project is not the same project of the bug we are
+			# viewing, override the current project. This to avoid problems with
+			# categories and handlers lists etc.
+			global $g_project_override;
+			$g_project_override = $this->sourceIssue->project_id;
+		}
```

Comment:
I see this block called multiple times. Consider refactoring out.

Reply:
Yes, I'm planning to do some horizontal refactoring across commands, but I'm trying to get the commands checked in first, before doing refactoring that would conflict with others.